### ISSUE TITLE
I think my data type is wrong when doing video_transform(video)

### ISSUE THREAD

**Issue by PHDPeter** [ekosman/AnomalyDetectionCVPR2018-Pytorch#14]
Hi
So when trying to run feature_extractor.py
I am running into issues inside of self.video_transform() more particularly in the Resize(Transform) class of image_transfroms.py (around line 67). 
The first issues I had was that data.shape was outputting 4 variable rather than 3. 
this is an easy fix but when the code goes on to cv2.resize() it becomes an issue.

Looking into what exactly the "data" variable is made up of I found it was a   <class 'torch.Tensor'> and the with and hight (the first 2 variables coming out of data.shape) were 256 and 256 respectively.

This does not look right to me as cv2.resize() is normally meant for an image rather than a <class 'torch.Tensor'>  object, right?

Could someone with working code just let me know if the data variable at this point in the code should be of type <class 'torch.Tensor'> ?

Thanks in advance.

**Comment by PHDPeter** [comment]
sorry,
with and hight are 240 and 35
the new with and hight were 240 by 240.
Sorry

**Comment by PHDPeter** [comment]
Using openCV 3.4.2 and pytorch 1.4.0

**Comment by PHDPeter** [comment]
Ok looking further into it I found that the output of self.video_clips.get_clip(idx) should be a tensor so that is alright but I am still struggling to find why its shape is 4 long rather than 3?
Does anyone else get a tensor that is 4 long for the data or video variable?

**Comment by PHDPeter** [comment]
@ekosman I was looking into this problem and you seem to have posted about this kind of problem before on a different project (https://github.com/pytorch/vision/issues/1338) as such I am hoping you might have some insight into why I am getting this error?
I am using a slightly higher torch and torchvision version then you do in the project but I don't think this is the issues as I have tried with the lower version and still got this error.
Thanks,

**Comment by PHDPeter** [comment]
Ok, I have tracked the problem down.
It would seem that I am getting all the 0 in my tensor as the read_video() function of video.py is having trouble converting the 'yuv420p' format of the frame into "rgb", vai the pyav function to_rgb()
As this is not a problem already reported on here, I have a feeling it is an issue with my set up.
I have been using av==6.2.0 (the same as in the environment.yml) and my FFmgp == 4.2.2 so I don't think that is the issue.
Does anyone have any Ideas where I am going wrong?
I know this is not really to do with this git but any help would be appreciated :-)

**Comment by PHDPeter** [comment]
I have tracked the problem down to Pyav (https://github.com/mikeboers/PyAV/issues/602) 
thanks.

**Comment by ekosman** [comment]
@PHDPeter Seems like the transforms file included here is not the right file. I will fix this once I have more time to get into this. Basically, it should use the original video transforms from torchvision,

**Comment by ekosman** [comment]
@PHDPeter I've updated the code, please let me know if it works

**Comment by PHDPeter** [comment]
@ekosman Thanks.
Thats seems to have fixted the probelm with tranfroming the image, Thank you.
Thou now I get a error:
"index 4 is out of bounds for dimension 0 with size 3"
I assume this is coming from the fact "video" is now a tensor with shape 3 so the index is now not working as it was expecting the a array storing rgb for each frame?
can this be sloved by just covering the tensor back in to a np.array?
would av.to_nparray() work?? I am not sure what shape you wanted video to be by that ponit, as my code did not get that fair before :-)

**Comment by PHDPeter** [comment]
@ekosman I.e. should the video of the getitem_from_raw_video() be a ndarry of shape (3,32,224,224)?
If so I have fixed it, if not I have not..

**Comment by ekosman** [comment]
@PHDPeter `video` should be a 4-dim tensor. An image is a 3-dim tensor, thus a video clip which is a stack of images should have one more dim - 4

**Comment by PHDPeter** [comment]
Ow know you are right it is 4 dim tensor
 `test=video.shape
4`
sorry about that

But what is the video[in_clip_frames] is meant to be doing? As the in_clip_frames is a list with value that are much larger then the index video could support i.e.  in_clip_frames-[0,2,4,6,6,10,12,14,16,18,20,22,24,26,28,30]  but the index of video can't be above 4?

**Comment by ekosman** [comment]
@PHDPeter Well, I remember reading in the paper that the sampling of frames is "every 2 frames", probably because 2 subsequent frames look the same. Torchvision extracts gives me a 32-length video clips, and I pull out every second frame.

**Comment by PHDPeter** [comment]
I think the dimensions are right if you wanted it as a video
`video.shape)`
> `torch.Size([3,` 32, 224, 224])`  

But I have the feeling, from what the code used to look like, that it is meant to be dealing with "frames" of the video meaning they could be looking for them to be in 3 dim shape as they are images?

**Comment by ekosman** [comment]
c3d expects video clips, thus deals with 4-dim tensors (actually 5-dim tensors, when considering the batch dimension)

Thinking about it, I'll fix the dimensions problem now. The frames extraction shouldn't be `video[in_clip_frames]`, but something like `video[:, in_clip_frames, :, :]`.

I'll check it.

**Comment by PHDPeter** [comment]
Hi @ekosman  
sorry for the delay in replying.
As far as I can tell that fixed it.
`video[:, in_clip_frames, :, :]`
The feature extractor is working for me, thou I might be having some issues with the memory need to run the C3D over the whole list of training data.
But I am fairly sure that is unrelated 
Thanks
